x-pack/plugins/enterprise_search/public/applications/app_search/components/document_creation/document_creation_logic.ts

```diff
@@ -74,5 +99,122 @@ export const DocumentCreationLogic = kea<
         setFileInput: (_, { fileInput }) => fileInput,
       },
     ],
+    isUploading: [
+      false,
+      {
+        onSubmitFile: () => true,
+        onSubmitJson: () => true,
+        setErrors: () => false,
+        setSummary: () => false,
+      },
+    ],
+    warnings: [
+      [],
+      {
+        onSubmitJson: () => [],
+        setWarnings: (_, { warnings }) => warnings,
+      },
+    ],
+    errors: [
+      [],
+      {
+        onSubmitJson: () => [],
+        setErrors: (_, { errors }) => (Array.isArray(errors) ? errors : [errors]),
+      },
+    ],
+    summary: [
+      {} as DocumentCreationSummary,
+      {
+        setSummary: (_, { summary }) => summary,
+      },
+    ],
+  }),
+  listeners: ({ values, actions }) => ({
+    onSubmitFile: async () => {
+      const { fileInput } = values;
+
+      if (!fileInput) {
+        return actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_FILE]);
+      }
+      try {
+        const textInput = await readUploadedFileAsText(fileInput);
+        actions.setTextInput(textInput);
+        actions.onSubmitJson();
+      } catch {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NO_VALID_FILE]);
+      }
+    },
+    onSubmitJson: () => {
+      const { textInput } = values;
+
+      const MAX_UPLOAD_BYTES = 50 * 1000000; // 50 MB
+      if (Buffer.byteLength(textInput) > MAX_UPLOAD_BYTES) {
+        actions.setWarnings([DOCUMENT_CREATION_WARNINGS.LARGE_FILE]);
+      }
+
+      let documents;
+      try {
+        documents = JSON.parse(textInput);
+      } catch (error) {
+        return actions.setErrors([error.message]);
+      }
+
+      if (Array.isArray(documents)) {
+        actions.uploadDocuments({ documents });
+      } else if (isPlainObject(documents)) {
+        actions.uploadDocuments({ documents: [documents] });
+      } else {
+        actions.setErrors([DOCUMENT_CREATION_ERRORS.NOT_VALID]);
+      }
+    },
+    uploadDocuments: async ({ documents }) => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      const CHUNK_SIZE = 100;
+      const MAX_EXAMPLES = 5;
+
+      const promises = chunk(documents, CHUNK_SIZE).map((documentsChunk) => {
+        const body = JSON.stringify({ documents: documentsChunk });
+        return http.post(`/api/app_search/engines/${engineName}/documents`, { body });
+      });
+
+      try {
+        const responses = await Promise.all(promises);
+        const summary: DocumentCreationSummary = {
```

Comment:
This is not a change request. I wonder if a `summary` would have been better as a selector. Store the raw `response`, and use a `summary` selector to parse up the response into a summary. Just something to break up this monster logic file a bit. The `parseSummary` could even be its own function that could be unit tested independently.

Instead of a monster summary you could even have more fine grained selectors that parse up the response as needed.

Reply:
Super interesting, I really like that idea! I think my only hangup there is storing `response` in state, just because that feels a little weird since we're not really using it directly. I definitely agree the merging logic/`parseSummary` does feel like it could be a separate utility for sure.

Reply:
Yeah, something to think about. State doesn't necessarily need to be used directly.